packages/react-components/react-color-picker-preview/library/src/components/ColorArea/useColorArea.ts

```diff
@@ -9,23 +16,167 @@ import type { ColorAreaProps, ColorAreaState } from './ColorArea.types';
  * before being passed to renderColorArea_unstable.
  *
  * @param props - props from this instance of ColorArea
- * @param ref - reference to root HTMLDivElement of ColorArea
+ * @param ref - reference to root HTMLInputElement of ColorArea
  */
 export const useColorArea_unstable = (props: ColorAreaProps, ref: React.Ref<HTMLDivElement>): ColorAreaState => {
-  return {
-    // TODO add appropriate props/defaults
+  const { targetDocument } = useFluent();
+  const targetWindow = targetDocument?.defaultView!;
+  const rootRef = React.useRef<HTMLDivElement>(null);
+  const xRef = React.useRef<HTMLInputElement>(null);
+  const yRef = React.useRef<HTMLInputElement>(null);
+
+  const {
+    onChange,
+    // Slots
+    inputX,
+    inputY,
+    thumb,
+
+    ...rest
+  } = props;
+
+  const hsvColor = React.useMemo(() => props.color || undefined, [props.color]);
+  const defaultHsv = React.useMemo(() => props.defaultColor || undefined, [props.defaultColor]);
+  const [color, setColor] = useControllableState<HsvColor>({
+    defaultState: defaultHsv,
+    state: hsvColor,
+    initialState: INITIAL_COLOR_HSV,
+  });
+
+  const saturation = color.s < 1 ? color.s * 100 : color.s;
+  const value = color.v < 1 ? color.v * 100 : color.v;
+
+  const coordinates = { x: saturation, y: value };
+
+  function getCoordinates(event: MouseEvent) {
+    const rect = rootRef.current?.getBoundingClientRect();
+    if (!rect) {
+      return coordinates;
+    }
+    const newX = Math.round(((event.clientX - rect.left) / rect.width) * 100);
+    const newY = 100 - Math.round(((event.clientY - rect.top) / rect.height) * 100);
+
+    return {
+      x: clamp(newX, MIN, MAX),
+      y: clamp(newY, MIN, MAX),
+    };
+  }
+
+  const abortController = React.useRef(new AbortController()).current;
+
+  React.useEffect(() => {
+    return () => {
+      abortController.abort();
+    };
+  }, [abortController]);
+
+  const dispatchCustomInputChangeEvent = (newColor: HsvColor) => {
+    const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
+      targetWindow.HTMLInputElement.prototype,
+      'value',
+    )!.set;
+    nativeInputValueSetter?.call(yRef.current!, newColor.v.toString());
+    nativeInputValueSetter?.call(xRef.current!, newColor.s.toString());
+
+    const event = new Event('change', { bubbles: true });
+
+    xRef.current?.dispatchEvent(event);
+    yRef.current?.dispatchEvent(event);
```

Comment:
Does this trigger two onChange callback calls?

Reply:
That part is already removed